### PR TITLE
(#310)CancerTypeHomePage-FrontEndVerification

### LINKED ISSUE #310: CMS: Cancer Type Homepage - front end verification and editing of content
## As a [role], I want [feature], so that [benefit]

> **ESTIMATE** TBD

### Acceptance Criteria

```gherkin
Scenario: Verify newly created content
        Given user is navigating to the front end site with path site section plus "cancer-type-homepage"
        And page title is "Automated Test Cancer Type Homepage Patient"
        And the following cards are displayed
            | title                    | cardType       |
            | Test Overview            | cthp-overview  |
            | Test Treatment           | cthp-treatment |
            | Test Causes & Prevention | cthp-causes    |
            | Test Statistics          | cthp-survival  |
            | Test Video Card          | cthp-screening |
            | Test Research            | cthp-research  |
            | Test Block Card          | cthp-general   |
            | Test Raw HTML Card       | cthp-genetics  |
        And "cthp-overview" card has description "Description of Overview card of CTHP content type"
        And PDQ link label reads "PDQ Treatment Information for Patients"
        Then the PDQ link is matching the earlier selected PDQ link
        When user clicks on "View more information" dropdown
        Then the following more info links are displayed
            | title                                 | url                           |
            | Article to test Related Resources     | {TEST_SITE_SECTION}/article   |
            | Google Link Guide Card                | https://www.google.com        |
            | Media Link Override Title Guide cards | {TEST_SITE_SECTION}/test-file |
        And "cthp-causes" card has a link "Article to test Related Resources" with href "{TEST_SITE_SECTION}/article"
        And "cthp-survival" card has a link "Override Card Desc - External Feature Card" with href "https://www.google1.com"
        Then the video is matching the earlier selected video
        And "cthp-screening" card has a link "Article to test Related Resources" with href "{TEST_SITE_SECTION}/article"
        And "cthp-general" card has description that is not empty
        And "cthp-general" card has multiple links
        And "cthp-genetics" card reads "This is to test Raw Html Content in CTHP Raw HTML Card"
        And "cthp-research" card has multiple links
        Then user clicks "View more research" link inside "cthp-research" card
        Then the Cancer Research List page title is matching the earlier selected Cancer Research List page title

    Scenario: Edit and republish Cancer Type Homepage content type
        Given user is navigating to "/user/login"
        When user enters credentials
        And user clicks "Log in" button
        Then user is logged in and the user name "admin" is displayed in the toolbar
        And the tool bar appears at the top
        When user clicks on "Content" tab
        And user clicks on title with url "cancer-type-homepage" from the list of content
        And user clicks on the tool bar status green button "Published"
        And user clicks "View in edit form" button from other actions
        And user clears out "Pretty URL" field
        And user clears out "Meta Description" field
        And user fills out the following fields
            | fieldLabel       | value                                                       | field_name                                                                                          |
            | Pretty URL       | cancer-type-homepage-edited                                 | field_pretty_url                                                                                    |
            | Meta Description | Automated Test Cancer Type Homepage Meta Description Edited | field_page_description                                                                              |  |
            | CTHP Card Title  | Edited                                                      | field_cthp_cards[0][subform][field_cthp_card_title]                                                 |
            | CTHP Card Title  | Edited                                                      | field_cthp_cards[1][subform][field_cthp_card_title]                                                 |
            | PDQ Link Heading | Edited                                                      | field_cthp_cards[1][subform][field_cthp_pdq_link_heading][0][value]                                 |
            | Override Title   | Edited                                                      | field_cthp_cards[1][subform][field_cthp_view_more_info][2][subform][field_override_title][0][value] |
        And user removes Video from "CTHP Video card" card area
        And user clicks on "Select video" to choose a resource to link
        And user selects 2 Video from the list of videos
        And user clicks on "Select media" button to select media
        And user remembers the title of selected video for further verification
        When user saves the content page
        And user clicks on the tool bar status green button "Editing"
        And user selects "Quick Publish" from workflow actions

    Scenario: Verify edited content
        Given user is navigating to the front end site with path site section plus "cancer-type-homepage-edited"
        Then page title is "Automated Test Cancer Type Homepage Patient Edited"
        And the page contains meta tags with the following names
            | name        | content                                                     |
            | description | Automated Test Cancer Type Homepage Meta Description Edited |
        And the following cards are displayed
            | title                | cardType       |
            | Test OverviewEdited  | cthp-overview  |
            | Test TreatmentEdited | cthp-treatment |
        And "cthp-overview" card has description "Description of Overview card of CTHP content type"
        And PDQ link label reads "PDQ Treatment Information for PatientsEdited"
        When user clicks on "View more information" dropdown
        Then the following more info links are displayed
            | title                                       | url                           |
            | Article to test Related Resources           | {TEST_SITE_SECTION}/article   |
            | Google Link Guide Card                      | https://www.google.com        |
            | Media Link Override Title Guide cardsEdited | {TEST_SITE_SECTION}/test-file |
        Then the video is matching the earlier selected video
      
    Scenario: Clean up
        Given user is navigating to "/user/login"
        When user enters credentials
        And user clicks "Log in" button
        Then user is logged in and the user name "admin" is displayed in the toolbar
        And the tool bar appears at the top
        When user clicks on "Content" tab
        And user selects a checkbox next to title with url "cancer-type-homepage-edited" from the list of content
        And user clicks on "Apply to selected items" action button
        Then the page title is "Are you sure you want to delete this content item?"
        When user clicks on "Delete" button
        Then the text "Deleted 1 content item." appears on a screen
        And the content item with url "cancer-type-homepage-edited" does not exist in the list of content
```

### Resources:
* Feature: JIRA Ticket #
* Mockups: *Here goes a URL to or the name of the mockup(s) in inVision*
* Web Page: *Here goes a URL to an existing web page*

### Notes

*Some complementary notes if necessary:*

* > Here goes a quote from an email
* Here goes whatever useful information can exist…

---

### Prerequisites
- *required dependency for identifying the problem, scope and then solution for this issue*
- [ ] Short Spike Description - Issue #9999

### Tasks
- [ ] Short Task Description - Issue #9999
- [ ] Short Task Description - Issue #9999

### Technical Notes
- More notes, or technical nuances/direction

### PR DESCRIPTION
**Pull request by ahmadakbaryNIH**
closes #310

### CODE REVIEW

**Comment by ahmadakbaryNIH** [comment]
image

**Comment by ahmadakbaryNIH** [comment]
image

**Comment by ahmadakbaryNIH** [comment]
image
image
image
image

**Comment by ahmadakbaryNIH** [comment]
$CancerTypeHomePagefrontEndVerification
image

File: cypress/e2e/CMS/$CancerTypeHomePage/index.js [hunk 1]
```diff
@@ -163,5 +140,121 @@ And('user selects {string} from draft {string} dropdown', (dropdown, option) =>
...
+    for (const { title, url } of dataTable.hashes()) {
+        const replacedTestSiteSection = url.replace("{TEST_SITE_SECTION}", siteSection);
+        cy.get(`ul a[class='title']:contains("${title}")`).should('be.visible').and('have.attr', 'href', replacedTestSiteSection)
+    }
+})
+
+And('cthp causes card has a link {string} with href {string}', (linkText, href) => {
+    const replacedTestSiteSection = href.replace("{TEST_SITE_SECTION}", siteSection);
+    cy.get("div[class='equalheight bgWhite cthp-causes'] a").should('have.text', linkText)
+    cy.get("div[class='equalheight bgWhite cthp-causes'] a").should('have.attr', 'href', replacedTestSiteSection)
```
Comment (KateMashkinaNIH) [inline cypress/e2e/CMS/$CancerTypeHomePage/index.js:213]: the locator can be changed to div[class*='cthp-causes'] a
also, make use of aliases so you do not repeat your code
cy.get("div[class='equalheight bgWhite cthp-causes'] a").as('card').should('have.text', linkText)
    cy.get('@ card').should('have.attr', 'href', replacedTestSiteSection)

File: cypress/e2e/CMS/$CancerTypeHomePage/index.js [hunk 2]
```diff
@@ -163,5 +140,121 @@ And('user selects {string} from draft {string} dropdown', (dropdown, option) =>
...
+})
+
+And('cthp survival card has a link {string} with href {string}', (title, href) => {
+    cy.get(`div[class='equalheight bgWhite cthp-survival'] a:contains("${title}")`).should('be.visible').and('have.attr', 'href', href)
```
Comment (KateMashkinaNIH) [inline cypress/e2e/CMS/$CancerTypeHomePage/index.js:217]: this should be handled like a step above ( also, update class locator how I suggested above a well) 
cy.get("").as('card').should('have.text', linkText)
cy.get('@ card').should('have.attr', 'href', replacedTestSiteSection)

Comment (KateMashkinaNIH) [reply cypress/e2e/CMS/$CancerTypeHomePage/index.js:217]: no space after @

File: cypress/e2e/CMS/$CancerTypeHomePage/index.js [hunk 3]
```diff
@@ -163,5 +140,121 @@ And('user selects {string} from draft {string} dropdown', (dropdown, option) =>
...
+})
+
+And('cthp screening card has a link {string} with href {string}', (linkText, href) => {
+    const replacedTestSiteSection = href.replace("{TEST_SITE_SECTION}", siteSection);
+    cy.get("div[class='equalheight bgWhite cthp-screening'] a").should('have.text', linkText)
```
Comment (KateMashkinaNIH) [inline cypress/e2e/CMS/$CancerTypeHomePage/index.js:222]: pls update locator 
and do the same I suggested for two prev steps

File: cypress/e2e/CMS/$CancerTypeHomePage/index.js [hunk 4]
```diff
@@ -163,5 +140,121 @@ And('user selects {string} from draft {string} dropdown', (dropdown, option) =>
...
+    cy.get("div[class='equalheight bgWhite cthp-screening'] a").should('have.attr', 'href', replacedTestSiteSection)
+})
+
+And('cthp general card has description that is not empty', () => {
+    cy.get("div[class='equalheight bgWhite cthp-general'] p").should('be.visible')
```
Comment (KateMashkinaNIH) [inline cypress/e2e/CMS/$CancerTypeHomePage/index.js:227]: update locator
also, just checking that it is visible is not enough - you should add that it is not empty

File: cypress/e2e/CMS/$CancerTypeHomePage/index.js [hunk 5]
```diff
@@ -163,5 +140,121 @@ And('user selects {string} from draft {string} dropdown', (dropdown, option) =>
...
+})
+And('cthp general card has multiple links', () => {
+    cy.get("div[class='equalheight bgWhite cthp-general'] p").should('be.visible')
+})
```
Comment (KateMashkinaNIH) [inline cypress/e2e/CMS/$CancerTypeHomePage/index.js:231]: this does not check for links at all 
you should be checking that div[class*='cthp-general'] a ).length is more than 0

File: cypress/e2e/CMS/$CancerTypeHomePage/index.js [hunk 6]
```diff
@@ -163,5 +140,121 @@ And('user selects {string} from draft {string} dropdown', (dropdown, option) =>
...
+
+And('cthp genetics card reads {string}', (title) => {
+    cy.get(`div[class='equalheight bgWhite cthp-genetics'] div:contains("${title}")`).should('be.visible')
+})
+
+And('cthp research card has multiple links', () => {
+    cy.get(`div[class='equalheight bgWhite cthp-research'] a`).should('be.visible')
```
Comment (KateMashkinaNIH) [inline cypress/e2e/CMS/$CancerTypeHomePage/index.js:238]: same as similar step cthp general card has multiple links

File: cypress/e2e/CMS/$CancerTypeHomePage/index.js [hunk 7]
```diff
@@ -163,5 +140,121 @@ And('user selects {string} from draft {string} dropdown', (dropdown, option) =>
...
+})
+
+And('user clicks {string} link inside cthp research card', (option) => {
+    cy.get(`a:contains("${option}")`).click({ force: true })
```
Comment (KateMashkinaNIH) [inline cypress/e2e/CMS/$CancerTypeHomePage/index.js:242]: this should be clicking a link inside research card and your locator should reflect that 
div[class*='cthp-research'] a:contains("${option}")

File: cypress/e2e/CMS/$CancerTypeHomePage/index.js [hunk 8]
```diff
@@ -163,5 +140,121 @@ And('user selects {string} from draft {string} dropdown', (dropdown, option) =>
...
+})
+
+And('user removes Video from {string} card area', (section) => {
+    cy.get(`div:contains("${section}")`).parent().find("input[name*='field_cthp_video_remove']").click({ force: true })
+})
+
+And('user selects {int} Video from the list of videos', (num) => {
+    cy.getIframeBody('iframe.entity-browser-modal-iframe').find('input.form-checkbox').first().eq(num - 1).check()
```
Comment (KateMashkinaNIH) [inline cypress/e2e/CMS/$CancerTypeHomePage/index.js:250]: this is redundunt 
.first().eq(num - 1) - first() grabs the first element, and then you pass an index - I think you need to remove first here

**Review by KateMashkinaNIH — CHANGES_REQUESTED** [review]

**Comment by ahmadakbaryNIH** [comment]
image

File: cypress/e2e/CMS/$CancerTypeHomePage/index.js [hunk 9]
```diff
@@ -163,5 +140,120 @@ And('user selects {string} from draft {string} dropdown', (dropdown, option) =>
...
+    for (const { title, url } of dataTable.hashes()) {
+        const replacedTestSiteSection = url.replace("{TEST_SITE_SECTION}", siteSection);
+        cy.get(`ul a[class='title']:contains("${title}")`).should('be.visible').and('have.attr', 'href', replacedTestSiteSection)
+    }
+})
+
+And('cthp causes card has a link {string} with href {string}', (linkText, href) => {
+    const replacedTestSiteSection = href.replace("{TEST_SITE_SECTION}", siteSection);
+    cy.get("div[class='equalheight bgWhite cthp-causes'] a").as('card').should('have.text', linkText)
+    cy.get("div[class*='cthp-causes'] a").as('card').should('have.attr', 'href', replacedTestSiteSection)
```
Comment (KateMashkinaNIH) [inline cypress/e2e/CMS/$CancerTypeHomePage/index.js:212]: so you used alias in the first line, then why are you repeating your code in the second? change it this
cy.get("div[class*='cthp-causes'] a").as('card').should('have.text', linkText)
    cy.get("@card").should('have.attr', 'href', replacedTestSiteSection)

File: cypress/e2e/CMS/$CancerTypeHomePage/index.js (hunk 2 above)
Comment (KateMashkinaNIH) [reply cypress/e2e/CMS/$CancerTypeHomePage/index.js:216]: change it to 
cy.get("div[class*='cthp-survival'] a").as('card').should('have.text', linkText)
cy.get("@card").should('have.attr', 'href', replacedTestSiteSection)

File: cypress/e2e/CMS/$CancerTypeHomePage/index.js [hunk 10]
```diff
@@ -163,5 +140,120 @@ And('user selects {string} from draft {string} dropdown', (dropdown, option) =>
...
+})
+
+And('cthp survival card has a link {string} with href {string}', (title, href) => {
+    cy.get(`div[class='equalheight bgWhite cthp-survival'] a:contains("${title}")`).should('be.visible').and('have.attr', 'href', href)
+})
 
+And('cthp screening card has a link {string} with href {string}', (linkText, href) => {
+    const replacedTestSiteSection = href.replace("{TEST_SITE_SECTION}", siteSection);
+    cy.get("div[class='equalheight bgWhite cthp-screening'] a").as('card').should('have.text', linkText)
+    cy.get("div[class*='cthp-screening'] a").as('card').should('have.attr', 'href', replacedTestSiteSection)
```
Comment (KateMashkinaNIH) [inline cypress/e2e/CMS/$CancerTypeHomePage/index.js:222]: cy.get("div[class*='cthp-screening'] a").as('card').should('have.text', linkText)
cy.get("@card").should('have.attr', 'href', replacedTestSiteSection)

**Review by KateMashkinaNIH — CHANGES_REQUESTED** [review]

**Comment by ahmadakbaryNIH** [comment]
image

**Comment by ahmadakbaryNIH** [comment]
image

**Review by KateMashkinaNIH — APPROVED** [review]

**Pull request merged**